test/jdk/java/foreign/TestBufferStack.java

```diff
@@ -0,0 +1,122 @@
+                segment2 = frame2.allocate(JAVA_LONG);
+                Assert.assertEquals(segment2.address(), segment12.address() + /*segment12 size + frame 1 spare + alignment constraint*/ 3 * JAVA_INT.byteSize());
+                Assert.assertEquals(segment2.byteSize(), JAVA_LONG.byteSize());
+                Assert.assertEquals(segment2.scope(), frame2.scope());
+                segment2.set(JAVA_LONG, 0, 1);
+
+                // Frames must be closed in stack order.
+                Assert.assertThrows(IllegalStateException.class, frame1::close);
+            }
+            // Scope is closed here, inner segments throw.
+            Assert.assertThrows(IllegalStateException.class, () -> segment2.get(JAVA_INT, 0));
+            // A new stack frame allocates at the same location (but different scope) as the previous did.
+            try (Arena frame3 = stack.pushFrame(2 * JAVA_INT.byteSize(), JAVA_INT.byteAlignment())) {
+                MemorySegment segment3 = frame3.allocate(JAVA_INT);
+                Assert.assertEquals(segment3.scope(), frame3.scope());
+                Assert.assertEquals(segment3.address(), segment12.address() + 2 * JAVA_INT.byteSize());
+            }
+
+            // Fallback arena behaves like regular stack frame.
+            MemorySegment outOfStack;
+            try (Arena hugeFrame = stack.pushFrame(1024, 4)) {
+                outOfStack = hugeFrame.allocate(4);
+                Assert.assertEquals(outOfStack.scope(), hugeFrame.scope());
+                Assert.assertTrue(outOfStack.asOverlappingSlice(stackSegment).isEmpty());
+            }
+            Assert.assertThrows(IllegalStateException.class, () -> outOfStack.get(JAVA_INT, 0));
+
+            // Outer segments are still accessible.
+            segment11.get(JAVA_INT, 0);
+            segment12.get(JAVA_INT, 0);
+        }
+    }
+
+    @Test
+    public void stress() throws InterruptedException {
+        BufferStack stack = new BufferStack(256);
+        Thread[] vThreads = IntStream.range(0, 1024).mapToObj(_ ->
+                Thread.ofVirtual().start(() -> {
+                    long threadId = Thread.currentThread().threadId();
+                    while (true) {
+                        try (Arena arena = stack.pushFrame(JAVA_LONG.byteSize(), JAVA_LONG.byteAlignment())) {
+                            // Try to assert no two vThreads get allocated the same stack space.
+                            MemorySegment segment = arena.allocate(JAVA_LONG);
+                            JAVA_LONG.varHandle().setVolatile(segment, 0L, threadId);
+                            Assert.assertEquals(threadId, (long) JAVA_LONG.varHandle().getVolatile(segment, 0L));
+                        }
+                    }
+                })).toArray(Thread[]::new);
+        Thread.sleep(Duration.of(10, SECONDS));
+        Arrays.stream(vThreads).forEach(
+                thread -> Assert.assertTrue(thread.isAlive()));
+    }
+}
```

Comment:
I think we should also have a test that exercises the new code in the context of the linker. We have some tests that test downcall -> upcall with by-value structs, but I think we should also have a test that keeps going until it exhausts the buffer. Something like `test/jdk/java/foreign/stackwalk/TestReentrantUpcalls.java` (but without the `Whitebox` stuff).

Reply:
Done.